src/modules/addimage.js

```diff
@@ -730,20 +734,28 @@ import { atob, btoa } from "../libs/AtobBtoa.js";
   var arrayBufferToBinaryString = (jsPDFAPI.__addimage__.arrayBufferToBinaryString = function(
     buffer
   ) {
-    try {
-      return atob(btoa(String.fromCharCode.apply(null, buffer)));
-    } catch (e) {
-      if (
-        typeof Uint8Array !== "undefined" &&
-        typeof Uint8Array.prototype.reduce !== "undefined"
-      ) {
-        return new Uint8Array(buffer)
-          .reduce(function(data, byte) {
-            return data.push(String.fromCharCode(byte)), data;
-          }, [])
-          .join("");
+    // Skip direct conversion as it might take many hundred milliseconds before throwing.
+    if (buffer.length < ARRAY_BUFFER_LIMIT) {
+      try {
+        return atob(btoa(String.fromCharCode.apply(null, buffer)));
+      } catch (e) {
+        // Buffer was overflown, fall back to other methods
       }
     }
+    // Text decoder is much faster than string concatenation or reduce/join.
+    if (typeof TextDecoder !== "undefined") {
+      var decoder = new TextDecoder("ascii");
+      return decoder.decode(buffer);
```

Comment:
Actually, for binary strings, this is not going to work. ASCII only covers 127 code points, i.e. 0x00 through 0x7F. We need 0x00 through 0xFF (the entire byte range). For example, this fails:

```js
const inputBuffer = new Uint8Array([252, 129, 187, 147]);
const binStr = new TextDecoder('ascii').decode(inputBuffer.buffer);
console.log(binStr.split('').map(str => str.charCodeAt(0)));
// [252, 129, 187, 8220]
```
As it turns out, TextDecoder doesn't work well for binary string encoding, and this case should probably be removed entirely.

Reply:
I'm not sure how this passed the test cases actually...Any ideas? Since that simple example yielded malformed output I would expect something to have broken.

Reply:
Not sure how it passed library tests but it passed my own visual validation because the color palette was so small. In the end I did find it produced malformed data as some grays were going to green.
For some reason I had in mind ascii was 8 bit as well. And after some research TextDecoder still isn't a choice with any given text encoding to replace fromCharCode...